### PR TITLE
Auto excludes

### PR DESCRIPTION
**Pull request by romanowski**
Needs #2 to be merged

Now we are able to autodetect ThreadLocal like methods and synchronoized-like methods

### CODE REVIEW

File: run.py [hunk 1]
```diff
@@ -74,16 +76,21 @@ def call_compiler(scalaLocation, output, additionalOptions=[]):
                     scalacOptions +
                     sources +
                     debugOptions +
+                    additionalScalacOptions +
                     options.additionalOptions)
+    return time.time() - timeBefore
 
 
-call_compiler(options.scala, scalaOutput)
+compilation_time = call_compiler(options.scala, scalaOutput, ["-Yparallel-phases:parser", "-Yparallel-threads", "16"])
```
Comment (pkukielka) [inline run.py]: Revert. We want to be able to test this with different phases, different amount of threads, sometimes without threads etc

File: umad/src/main/java/com/pkukielka/AccessMonitorRewriter.java [hunk 2]
```diff
@@ -51,10 +57,15 @@ public static void logUnsafeMethodCalls(String methodName, String ifCalledFrom,
             if (last != null &&
                     last.threadId != current.threadId &&
                     last.hashCode == current.hashCode &&
-                    current.timestamp - last.timestamp <= conf.intervalMs)
-            {
+                    current.timestamp - last.timestamp <= conf.intervalMs) {
                 StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
 
+                for(StackTraceElement elem: stackTrace){
+                    String fqn = elem.getClassName() + "." +elem.getMethodName();
```
Comment (pkukielka) [inline umad/src/main/java/com/pkukielka/AccessMonitorRewriter.java]: Formatting

File: umad/src/main/java/com/pkukielka/AppConfig.java [hunk 3]
```diff
@@ -8,6 +8,7 @@
     public int intervalMs = 100;
     public boolean shouldThrowExceptions = true;
     public boolean shouldPrintStackTrace = false;
+    public boolean shouldStoreSafeMethod = false;
```
Comment (pkukielka) [inline umad/src/main/java/com/pkukielka/AppConfig.java]: Why we need that?

File: umad/src/main/java/com/pkukielka/AccessMonitorRewriter.java [hunk 4]
```diff
@@ -79,15 +90,105 @@ public static void logUnsafeMethodCalls(String methodName, String ifCalledFrom,
...
+    }
+
+    private static Set<Integer> methodOpcodes = new HashSet<Integer>(Arrays.asList(
+            Opcode.INVOKEVIRTUAL,
+            Opcode.INVOKESPECIAL,
+            Opcode.INVOKEINTERFACE,
+            Opcode.INVOKESTATIC
+    ));
+
+    private boolean isThreadLocalGetOrSet(ConstPool pool, int index) {
```
Comment (pkukielka) [inline umad/src/main/java/com/pkukielka/AccessMonitorRewriter.java]: I see nothing specific to `ThreadLocal` inside that method. Consider renaming it?

File: umad/src/main/java/com/pkukielka/AppConfig.java (hunk 3 above)
Comment (romanowski) [reply umad/src/main/java/com/pkukielka/AppConfig.java]: Tests

Comment (pkukielka) [reply umad/src/main/java/com/pkukielka/AppConfig.java]: That is what I thought, and I do not like it :P 
Let me look at the code again...

Comment (pkukielka) [reply umad/src/main/java/com/pkukielka/AppConfig.java]: I think it's unneeded. You can just check output to see if method was reported or not. And it's better end-to-end test also.

**Pull request merged**